### DISCUSSION TITLE
400 response, missing model parameter

### DISCUSSION THREAD

**Discussion by chrisecoleman** [srcnalt/OpenAI-Unity#52] [Q&A] [answered]
Hello, im using the sample project, which works well in the editor, but when attempting in Webgl build i get the following:

    "error": {
        "message": "you must provide a model parameter",
        "type": "invalid_request_error",
        "param": null,
        "code": null
    }

Has anyone experienced this?  I do see the model being set in code:

            // Complete the instruction
            var completionResponse = await openai.CreateChatCompletion(new CreateChatCompletionRequest()
            {
                Model = "gpt-3.5-turbo-0301",
                Messages = messages
            });

Suggestions?

**Comment by srcnalt** [comment]
Hi @chrisecoleman, similar issues reported before that certain versions of Unity does not handle requests properly. You may try on different versions of Unity and see if it works fine.

**Comment by chrisecoleman** [comment]
Thanks for your prompt response, I’ll check it on a different version tonight.  I’m on the latest build, do you have a recommended version?

**Comment by chrisecoleman** [comment]
Unfortunately the issue persist, ive tried 4 different unity versions.  If there is a reccomended version i can target that.  I've tried a few different unity libraries for openai, but none are functional for me in webgl, this seems to be the closest.  appreciate any further advice you may have.

**Comment by srcnalt** [comment]
I was able to run chat completion built from 2022.2 
<img width="1280" alt="image" src="https://user-images.githubusercontent.com/3163281/235477381-2e2728d2-20d0-45a6-a8f5-365c1e679713.png">

**Comment by chrisecoleman** [comment]
![image](https://user-images.githubusercontent.com/19179921/235505415-e3674886-9d99-4635-a5ec-eabf7d6fe078.png)

Ok, that is encouraging, i tried a fresh build just now.
I created a new 2d project in unity 2022.2.17.
i went to package manager and imported the github link.
i also imported the chatgpt sample.
i then moved the gpt sample scene into my heirarchy, and unloaded the sample scene.
i then included my api key in the constructor in line 17 of chatgpt.cs:
private OpenAIApi openai = new OpenAIApi() to private OpenAIApi openai = new OpenAIApi("myactualkey"). i realize this is insecure, but i wanted a quick test.
I then ran within the editor successfully
i then went to build settings, changed to webgl, and included the gchatgpt scene in the build config.
I then build and run
browser loads, and i sent my chat, but get no response, and devtools shows a status 400 and a response indicating i need to provide the model.

are there steps i am doing differently than you?  I really appreciate your help with this!

**Reply by chrisecoleman** [reply]
tried 2022.2.8f1 as well, same result

**Reply by srcnalt** [reply]
In OpenAiAPI.cs just before making the request can you output the json payload and see if it looks correct? Also could you check which version of NewtonsoftJson is available in the project?

**Comment by chrisecoleman** [comment]
Here is the JSON in the **editor**:

`{"model":"gpt-3.5-turbo-0301","messages":[{"role":"user","content":"Act as a random stranger in a chat room and reply to the questions. Don't break character. Don't ever mention that you are an AI model.\nhello?"}],"temperature":1.0,"n":1,"stream":false,"presence_penalty":0.0,"frequency_penalty":0.0}`

When i run in the **browser**, the JSON is empty!

Here is where im calling it:

        private byte[] CreatePayload<T>(T request)
        {
            var json = JsonConvert.SerializeObject(request, jsonSerializerSettings);
                        Debug.Log("request = "+request);
                        Debug.Log("json = "+json);
            return Encoding.UTF8.GetBytes(json);
        }

the newtonsoft json version is 3.1.0

**Reply by chrisecoleman** [reply]
hey @srcnalt do you have any further suggestions here, still stuck on this here, and it does seem to be the serialization failing, and only in webgl.  if i publish to unity play it exhibits the same behavior.

Been looking for a unity OpenAI solution that works in webgl for a few weeks here, this feels closest :)

**Reply by srcnalt** [reply]
Hi @chrisecoleman, I'm back from a work trip this week and don't have much time. I'll try to take a look coming day, in the weekend at worst.

**Comment by chrisecoleman** [comment]
Can confirm this shows the same behavior when deployed to unity play instead of localhost.

**Comment by chrisecoleman** [comment, marked as answer]
I was able to work through this by removing the {Get; Set;} properties in the datatypes.cs file.  it seems newtonsoft has issues with serializign properties that are setup that way.

Thank you for your help!

**Reply by srcnalt** [reply]
Yep, I am also planning to sacrifice MS code convention to make fields work. For some reason Naming Strategies are not working in certain platforms when built. Glad you fixed it for yourself. NewtonSoft has now a new version again, i'll try that too maybe something could be improved.
